File: src/main/java/models/Activite.java
package models;

import java.time.LocalDateTime;

public class Activite {

    private int id;
    private String nom;
    private String description;
    private double prix;
    private String typeActivite;
    private double noteMoyenne;
    private int guideId;
    private int destinationId;
    private LocalDateTime dateDebut;
    private LocalDateTime dateFin;
    private String status;
    private Integer maxPlaces;
    private String image;

    public Activite() {}

    public Activite(String nom, String description, double prix, String typeActivite,
                    int guideId, int destinationId, LocalDateTime dateDebut, LocalDateTime dateFin, String status,
                    String image) {
        this.nom = nom;
        this.description = description;
        this.prix = prix;
        this.typeActivite = typeActivite;
        this.guideId = guideId;
        this.destinationId = destinationId;
        this.dateDebut = dateDebut;
        this.dateFin = dateFin;
        this.status = status;
        this.image = image;
    }

    public Activite(int id, String nom, String description, double prix, String typeActivite, double noteMoyenne,
                    int guideId, int destinationId, LocalDateTime dateDebut, LocalDateTime dateFin, String status,
                    String image) {
        this.id = id;
        this.nom = nom;
        this.description = description;
        this.prix = prix;
        this.typeActivite = typeActivite;
        this.noteMoyenne = noteMoyenne;
        this.guideId = guideId;
        this.destinationId = destinationId;
        this.dateDebut = dateDebut;
        this.dateFin = dateFin;
        this.status = status;
        this.image = image;
    }

    @Override
    public String toString() {
        return "Activite{" +
                "id=" + id +
                ", nom='" + nom + '\'' +
                ", description='" + description + '\'' +
                ", prix=" + prix +
                ", typeActivite='" + typeActivite + '\'' +
                ", noteMoyenne=" + noteMoyenne +
                ", guideId=" + guideId +
                ", destinationId=" + destinationId +
                ", dateDebut=" + dateDebut +
                ", dateFin=" + dateFin +
                ", status='" + status + '\'' +
                '}';
    }

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getNom() { return nom; }
    public void setNom(String nom) { this.nom = nom; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public double getPrix() { return prix; }
    public void setPrix(double prix) { this.prix = prix; }

    public String getTypeActivite() { return typeActivite; }
    public void setTypeActivite(String typeActivite) { this.typeActivite = typeActivite; }

    public double getNoteMoyenne() { return noteMoyenne; }
    public void setNoteMoyenne(double noteMoyenne) { this.noteMoyenne = noteMoyenne; }

    public int getGuideId() { return guideId; }
    public void setGuideId(int guideId) { this.guideId = guideId; }

    public int getDestinationId() { return destinationId; }
    public void setDestinationId(int destinationId) { this.destinationId = destinationId; }

    public LocalDateTime getDateDebut() { return dateDebut; }
    public void setDateDebut(LocalDateTime dateDebut) { this.dateDebut = dateDebut; }

    public LocalDateTime getDateFin() { return dateFin; }
    public void setDateFin(LocalDateTime dateFin) { this.dateFin = dateFin; }

    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public Integer getMaxPlaces() { return maxPlaces; }
    public void setMaxPlaces(Integer maxPlaces) { this.maxPlaces = maxPlaces; }
    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }
    private boolean flashSale;
    private Double flashPrice;
    private LocalDateTime flashExpiresAt;

    public boolean isFlashSale() { return flashSale; }
    public void setFlashSale(boolean flashSale) { this.flashSale = flashSale; }

    public Double getFlashPrice() { return flashPrice; }
    public void setFlashPrice(Double flashPrice) { this.flashPrice = flashPrice; }

    public LocalDateTime getFlashExpiresAt() { return flashExpiresAt; }
    public void setFlashExpiresAt(LocalDateTime flashExpiresAt) { this.flashExpiresAt = flashExpiresAt; }
}

File: src/Entity/Activite.php
<?php

namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity]
class Activite
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: "integer")]
    private int $id;

    #[ORM\Column(type: "string", length: 150, nullable: true)]
    #[Assert\NotBlank(message: "Le nom est obligatoire.")]
    private ?string $nom = '';

    #[ORM\Column(type: "text", nullable: true)]
    #[Assert\NotBlank(message: "La description est obligatoire.")]
    private ?string $description = '';

    #[ORM\Column(type: "float", nullable: true)]
    #[Assert\NotBlank(message: "Le prix est obligatoire.")]
    #[Assert\Positive(message: "Le prix doit être positif.")]
    private ?float $prix = 0;

    #[ORM\Column(name: "typeActivite", type: "string", length: 100, nullable: true)]
    #[Assert\NotBlank(message: "Le type d'activité est obligatoire.")]
    private ?string $typeActivite = '';

    #[ORM\Column(name: "noteMoyenne", type: "float", nullable: true)]
    private ?float $noteMoyenne = 0;

    #[ORM\ManyToOne(targetEntity: Guide::class, inversedBy: "activites")]
    #[ORM\JoinColumn(name: 'guide_id', referencedColumnName: 'id', onDelete: 'CASCADE')]
    private ?Guide $guide = null;

    #[ORM\Column(type: "datetime", nullable: true)]
    #[Assert\NotBlank(message: "La date de début est obligatoire.")]
    #[Assert\Type(\DateTimeInterface::class)]
    #[Assert\GreaterThanOrEqual(
        "today",
        message: "La date de début ne peut pas être antérieure à aujourd'hui."
    )]
    private ?\DateTimeInterface $date_debut = null;

    #[ORM\Column(type: "datetime", nullable: true)]
    #[Assert\NotBlank(message: "La date de fin est obligatoire.")]
    #[Assert\Type(\DateTimeInterface::class)]
    private ?\DateTimeInterface $date_fin = null;

    #[ORM\Column(type: "string", nullable: false, options: ["default" => "DISPONIBLE"])]
    private string $status = 'DISPONIBLE';

    #[ORM\ManyToOne(targetEntity: Ville::class, inversedBy: "activites")]
    #[ORM\JoinColumn(name: 'destination_id', referencedColumnName: 'id', onDelete: 'CASCADE')]
    #[Assert\NotNull(message: "La destination est obligatoire.")]
    private ?Ville $destination = null;

    #[ORM\Column(type: "integer", nullable: true)]
    #[Assert\NotBlank(message: "Le nombre maximum de places est obligatoire.")]
    #[Assert\PositiveOrZero(message: "Le nombre de places doit être positif ou nul.")]
    private ?int $max_places = 0;

    #[ORM\Column(type: "string", length: 255, nullable: true)]
    private ?string $image = '';

    #[ORM\Column(type: "boolean", nullable: false, options: ["default" => false])]
    private bool $is_flash_sale = false;

    #[ORM\Column(type: "float", nullable: true)]
    private ?float $flash_price = 0;

    #[ORM\Column(type: "datetime", nullable: true)]
    private ?\DateTimeInterface $flash_expires_at = null;

    #[ORM\OneToMany(mappedBy: "activite", targetEntity: Avis::class)]
    private Collection $aviss;

    #[ORM\OneToMany(mappedBy: "activite", targetEntity: Waitlist::class)]
    private Collection $waitlists;

    #[ORM\OneToMany(mappedBy: "activite", targetEntity: Ticket::class)]
    private Collection $tickets;

    #[ORM\OneToMany(mappedBy: "activite", targetEntity: Notification::class)]
    private Collection $notifications;

    public function __construct()
    {
        $this->aviss = new ArrayCollection();
        $this->waitlists = new ArrayCollection();
        $this->tickets = new ArrayCollection();
        $this->notifications = new ArrayCollection();

        $this->date_debut = new \DateTime();
        $this->date_fin = new \DateTime();
        $this->flash_expires_at = new \DateTime();
    }

    // --- Getters et Setters ---
    public function getId(): int { return $this->id; }

    public function getNom(): string { return $this->nom ?? ''; }
    public function setNom(?string $nom): self { $this->nom = $nom ?? ''; return $this; }

    public function getDescription(): string { return $this->description ?? ''; }
    public function setDescription(?string $description): self { $this->description = $description ?? ''; return $this; }

    public function getPrix(): float { return $this->prix ?? 0; }
    public function setPrix(?float $prix): self { $this->prix = $prix ?? 0; return $this; }

    public function getTypeActivite(): string { return $this->typeActivite ?? ''; }
    public function setTypeActivite(?string $typeActivite): self { $this->typeActivite = $typeActivite ?? ''; return $this; }

    public function getNoteMoyenne(): float { return $this->noteMoyenne ?? 0; }
    public function setNoteMoyenne(?float $noteMoyenne): self { $this->noteMoyenne = $noteMoyenne ?? 0; return $this; }

    public function getGuide(): ?Guide { return $this->guide; }
    public function setGuide(?Guide $guide): self { $this->guide = $guide; return $this; }

    public function getDateDebut(): \DateTimeInterface { return $this->date_debut ?? new \DateTime(); }
    public function setDateDebut(?\DateTimeInterface $date_debut): self { $this->date_debut = $date_debut ?? new \DateTime(); return $this; }

    public function getDateFin(): \DateTimeInterface { return $this->date_fin ?? new \DateTime(); }
    public function setDateFin(?\DateTimeInterface $date_fin): self { $this->date_fin = $date_fin ?? new \DateTime(); return $this; }

    public function getStatus(): string { return $this->status; }
    public function setStatus(string $status): self { $this->status = $status; return $this; }

    public function getDestination(): ?Ville { return $this->destination; }
    public function setDestination(?Ville $destination): self { $this->destination = $destination; return $this; }

    public function getMaxPlaces(): int { return $this->max_places ?? 0; }
    public function setMaxPlaces(?int $max_places): self { $this->max_places = $max_places ?? 0; return $this; }

    public function getImage(): string { return $this->image ?? ''; }
    public function setImage(?string $image): self { $this->image = $image ?? ''; return $this; }

    public function getIsFlashSale(): bool { return $this->is_flash_sale; }
    public function setIsFlashSale(bool $is_flash_sale): self { $this->is_flash_sale = $is_flash_sale; return $this; }

    public function getFlashPrice(): float { return $this->flash_price ?? 0; }
    public function setFlashPrice(?float $flash_price): self { $this->flash_price = $flash_price ?? 0; return $this; }

    public function getFlashExpiresAt(): ?\DateTimeInterface { return $this->flash_expires_at; }
    public function setFlashExpiresAt(?\DateTimeInterface $flash_expires_at): self { $this->flash_expires_at = $flash_expires_at ?? new \DateTime(); return $this; }

    // --- Collections ---
    public function getAviss(): Collection { return $this->aviss; }
    public function addAvis(Avis $avis): self {
        if (!$this->aviss->contains($avis)) {
            $this->aviss[] = $avis;
            $avis->setActivite($this);
        }
        return $this;
    }
    public function removeAvis(Avis $avis): self {
        if ($this->aviss->removeElement($avis)) {
            if ($avis->getActivite() === $this) { $avis->setActivite(null); }
        }
        return $this;
    }

    public function getWaitlists(): Collection { return $this->waitlists; }
    public function getTickets(): Collection { return $this->tickets; }
    public function getNotifications(): Collection { return $this->notifications; }
}